In [1]:
import pandas as pd
#import numpy as np
from sqlalchemy import create_engine
import warnings
import seaborn as sns

import matplotlib.pyplot as plt

%matplotlib inline

warnings.filterwarnings('ignore')

In [ ]:
#!pip install numpy==1.19.3

#As a workaround, you temporarily pin to NumPy 1.19.3 (`pip install numpy==1.19.3`) which has a fixed version of 
#OpenBLAS that works correctly in Windows 2004/20H2. 
#This fix broken other things (Linux, Windows-on-docker) and so could not be retained in NumPy 1.19.4.

In [2]:
import numpy as np

In [3]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from useducation',con=engine)

# No need for an open connection, 
# because you're only doing a single query
engine.dispose()

In [4]:
df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

          ...           GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  \
0         ...              57948.0     58025.0      41167.0      471564.0   
1         ...               9748.0      8789.0       6714.0       79117.0   
2         ...              55433.0     49081.0      37410.0      437127.0   
3         ...              34632.0     36011.0      27651.0      281338.0   
4         ...             418418.0    363296.0     270675.0     3286034.0   

   GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  \
0       196386.0      676174.0        208.327876        252.187522   
1        30847.0      112335.0               NaN               NaN   
2       175210.0      614881.0        215.253932        265.366278   
3       123113.0      405259.0        210.206028        256.312090   
4      1372011.0     4717112.0        208.398961        260.892247   

   AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0           207.963517                  NaN  
1                  NaN           258.859712  
2           206.212716           262.169895  
3           208.634458           264.619665  
4           196.764414                  NaN  

[5 rows x 25 columns]

1. Determine all the variable types and find the fraction of the missing values for each variable.
2. Notice that the data has a time dimension (year). For this assignment, forget about time and treat all the observations as if they're from the same year. Choose a strategy to deal with the missing values for each variable. For which variables would filling in the missing values with some value make sense? For which might tossing out the records entirely make sense?
3. Now, take into account the time factor. Replicate your second answer, but this time, fill in the missing values by using a statistic that is calculated within the year of the observation. For example, if you want to fill in a missing value for a variable with the mean of that variable, calculate the mean by using only the observations for that specific year.
4. This time, fill in the missing values using interpolation (extrapolation).
5. Compare your results for the second, third, and fourth questions. Do you find any meaningful differences?

In [5]:
# data types are 2 objects, 1 intergers and 22 floats
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1492 entries, 0 to 1491
Data columns (total 25 columns):
PRIMARY_KEY                     1492 non-null object
STATE                           1492 non-null object
YEAR                            1492 non-null int64
ENROLL                          1229 non-null float64
TOTAL_REVENUE                   1280 non-null float64
FEDERAL_REVENUE                 1280 non-null float64
STATE_REVENUE                   1280 non-null float64
LOCAL_REVENUE                   1280 non-null float64
TOTAL_EXPENDITURE               1280 non-null float64
INSTRUCTION_EXPENDITURE         1280 non-null float64
SUPPORT_SERVICES_EXPENDITURE    1280 non-null float64
OTHER_EXPENDITURE               1229 non-null float64
CAPITAL_OUTLAY_EXPENDITURE      1280 non-null float64
GRADES_PK_G                     1319 non-null float64
GRADES_KG_G                     1360 non-null float64
GRADES_4_G                      1361 non-null float64
GRADES_8_G                      1

In [6]:
df.isnull()

PRIMARY_KEY  STATE   YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0           False  False  False    True          False            False   
1           False  False  False    True          False            False   
2           False  False  False    True          False            False   
3           False  False  False    True          False            False   
4           False  False  False    True          False            False   
5           False  False  False    True          False            False   
6           False  False  False    True          False            False   
7           False  False  False    True          False            False   
8           False  False  False    True          False            False   
9           False  False  False    True          False            False   
10          False  False  False    True          False            False   
11          False  False  False    True          False            False   
12          False  False  False    True          False            False   
13          False  False  False    True          False            False   
14          False  False  False    True          False            False   
15          False  False  False    True          False            False   
16          False  False  False    True          False            False   
17          False  False  False    True          False            False   
18          False  False  False    True          False            False   
19          False  False  False    True          False            False   
20          False  False  False    True          False            False   
21          False  False  False    True          False            False   
22          False  False  False    True          False            False   
23          False  False  False    True          False            False   
24          False  False  False    True          False            False   
25          False  False  False    True          False            False   
26          False  False  False    True          False            False   
27          False  False  False    True          False            False   
28          False  False  False    True          False            False   
29          False  False  False    True          False            False   
...           ...    ...    ...     ...            ...              ...   
1462        False  False  False    True           True             True   
1463        False  False  False    True           True             True   
1464        False  False  False    True           True             True   
1465        False  False  False    True           True             True   
1466        False  False  False    True           True             True   
1467        False  False  False    True           True             True   
1468        False  False  False    True           True             True   
1469        False  False  False    True           True             True   
1470        False  False  False    True           True             True   
1471        False  False  False    True           True             True   
1472        False  False  False    True           True             True   
1473        False  False  False    True           True             True   
1474        False  False  False    True           True             True   
1475        False  False  False    True           True             True   
1476        False  False  False    True           True             True   
1477        False  False  False    True           True             True   
1478        False  False  False    True           True             True   
1479        False  False  False    True           True             True   
1480        False  False  False    True           True             True   
1481        False  False  False    True           True             True   
1482        False  False  False    True           True             True   
1483        False  False  False

In [7]:
df.describe()

YEAR        ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
count  1492.000000  1.229000e+03   1.280000e+03     1.280000e+03   
mean   2004.433646  9.159308e+05   9.092082e+06     7.663723e+05   
std       7.393983  1.065280e+06   1.174519e+07     1.145242e+06   
min    1992.000000  4.386600e+04   4.656500e+05     3.102000e+04   
25%    1998.000000  2.583140e+05   2.186305e+06     1.893540e+05   
50%    2004.000000  6.483130e+05   5.079546e+06     4.033765e+05   
75%    2011.000000  1.014528e+06   1.085985e+07     8.289660e+05   
max    2017.000000  6.307022e+06   8.921726e+07     9.990221e+06   

       STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
count   1.280000e+03   1.280000e+03       1.280000e+03   
mean    4.216553e+06   4.109157e+06       9.196681e+06   
std     5.543072e+06   5.482971e+06       1.197813e+07   
min     0.000000e+00   2.209300e+04       4.816650e+05   
25%     1.153097e+06   7.158345e+05       2.165404e+06   
50%     2.537074e+06   2.055780e+06       5.234506e+06   
75%     5.080939e+06   4.768680e+06       1.074519e+07   
max     5.090457e+07   3.610526e+07       8.532013e+07   

       INSTRUCTION_EXPENDITURE  SUPPORT_SERVICES_EXPENDITURE  \
count             1.280000e+03                  1.280000e+03   
mean              4.762966e+06                  2.680331e+06   
std               6.293004e+06                  3.353349e+06   
min               2.655490e+05                  1.399630e+05   
25%               1.168032e+06                  6.357900e+05   
50%               2.657452e+06                  1.525406e+06   
75%               5.568028e+06                  3.229651e+06   
max               4.396452e+07                  2.605802e+07   

       OTHER_EXPENDITURE         ...              GRADES_4_G     GRADES_8_G  \
count       1.229000e+03         ...             1361.000000    1361.000000   
mean        4.292046e+05         ...            64538.927259   64271.057311   
std         5.340693e+05         ...            80293.387625   78910.545776   
min         1.154100e+04         ...              633.000000     437.000000   
25%         1.028310e+05         ...            13739.000000   13342.000000   
50%         2.715960e+05         ...            43272.000000   43339.000000   
75%         5.186000e+05         ...            75481.000000   76566.000000   
max         3.995951e+06         ...           493415.000000  500143.000000   

        GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
count    1361.00000  1.361000e+03   1.361000e+03  1.319000e+03   
mean    54268.92432  5.192140e+05   2.470714e+05  8.024415e+05   
std     67889.27623  6.436384e+05   3.077056e+05  9.707031e+05   
min       311.00000  4.878000e+03   1.808000e+03  7.254000e+03   
25%     10919.00000  1.106260e+05   5.147100e+04  1.810650e+05   
50%     36735.00000  3.457750e+05   1.642600e+05  5.503420e+05   
75%     67460.00000  6.119050e+05   2.905020e+05  9.282755e+05   
max    498403.00000  3.929869e+06   2.013687e+06  5.944746e+06   

       AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
count        536.000000        532.000000           533.000000   
mean         234.768293        278.414711           218.866154   
std           10.221511         10.159722             7.769616   
min          187.134670        232.831510           178.557612   
25%          229.694352        272.761598           214.663401   
50%          237.238552        280.618803           220.416034   
75%          241.995486        285.347428           223.999337   
max          253.420961        300.568235           236.773867   

       AVG_READING_8_SCORE  
count           498.000000  
mean            263.661132  
std               6.803725  
min             236.379102  
25%             259.533171  
50%             265.010912  
75%             268.190121  
max             280.499130  

[8 rows x 23 columns]

In [8]:
# finding percentages of missing values
df.isnull().sum()*100/df.isnull().count()

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          17.627346
TOTAL_REVENUE                   14.209115
FEDERAL_REVENUE                 14.209115
STATE_REVENUE                   14.209115
LOCAL_REVENUE                   14.209115
TOTAL_EXPENDITURE               14.209115
INSTRUCTION_EXPENDITURE         14.209115
SUPPORT_SERVICES_EXPENDITURE    14.209115
OTHER_EXPENDITURE               17.627346
CAPITAL_OUTLAY_EXPENDITURE      14.209115
GRADES_PK_G                     11.595174
GRADES_KG_G                      8.847185
GRADES_4_G                       8.780161
GRADES_8_G                       8.780161
GRADES_12_G                      8.780161
GRADES_1_8_G                     8.780161
GRADES_9_12_G                    8.780161
GRADES_ALL_G                    11.595174
AVG_MATH_4_SCORE                64.075067
AVG_MATH_8_SCORE                64.343164
AVG_READING_4_SCORE             64

2. Notice that the data has a time dimension (year). For this assignment, forget about time and treat all the observations as if they're from the same year. Choose a strategy to deal with the missing values for each variable. For which variables would filling in the missing values with some value make sense? For which might tossing out the records entirely make sense?

ANSWER: The missing data is continuous and mostly in enrollment, the revenue columns and the average columns. If all are the same year, the we can impute the missing enrollments with the mean for the entire dataset. For the revenue and expenditure vars, it makes sense to impute missing values for each with the median since we're looking at money grouped by state because each is missing the same percentage of data, the percentage missing is somewhat small and we don't skew the data especially if we were analyzing revenue against expenditure. For the GRADES variable, the percent missing is even smaller. GRADES would likely be an important variable to analyze and we only have 1492 so we want to keep them and impute a mean by state.  

The remaining AVG variables have the highest percent missing and they are only missing every other grouping. I would impute the missing values by forward filling. 

In [ ]:
df.YEAR.unique()

In [9]:
df.YEAR.value_counts()

2004    59
2008    59
1999    59
2000    59
2001    59
2002    59
2003    59
2005    59
2007    59
2006    59
2009    59
2010    58
2011    58
2012    58
2013    58
1998    58
2015    57
1997    57
1996    57
1995    56
1994    56
1993    56
1992    56
2014    56
2016    51
2017    51
Name: YEAR, dtype: int64

In [10]:
df.STATE.value_counts()

DISTRICT_OF_COLUMBIA                        29
VIRGINIA                                    28
MAINE                                       26
CALIFORNIA                                  26
NEVADA                                      26
PENNSYLVANIA                                26
ILLINOIS                                    26
INDIANA                                     26
NEW_JERSEY                                  26
ARKANSAS                                    26
NEW_HAMPSHIRE                               26
ALABAMA                                     26
MINNESOTA                                   26
FLORIDA                                     26
MICHIGAN                                    26
NEW_YORK                                    26
ARIZONA                                     26
VERMONT                                     26
COLORADO                                    26
RHODE_ISLAND                                26
KENTUCKY                                    26
SOUTH_DAKOTA 

In [ ]:
df.head()

In [ ]:
df.sort_values(['YEAR','TOTAL_REVENUE'])

In [ ]:
df['STATE'].sort_values()

3. Now, take into account the time factor. Replicate your second answer, but this time, fill in the missing values by using a statistic that is calculated within the year of the observation. For example, if you want to fill in a missing value for a variable with the mean of that variable, calculate the mean by using only the observations for that specific year.

ANSWER: For ENROLL, we can impute a mean by year except for the first and last year which are missing all values for ENROLL. We can backfill 1992 and forward fill 2017.  

The revenue columns and the expenditure columns are missing mostly for certain areas in the US so we should remove areas in the US that contain mostly missing values (value counts < 23 because imputing a value would introduce too much error. Remaining nulls in these columns can be imputed with the median since we're dealing with money. OTHER EXPENDITURES column is null for 1992 so we can change those values to zero so that we can still analyze other columns for 1992 but keep in mind that this column can't be used. Revenue and expenditures for 2017 is missing so we should drop the year 2017.

The average columns are missing every other year so we impute the mean for all years since we are missing so much data. We could also choose to drop these missing values and limit our analysis to every other year but I'll choose use the overall mean so that we can still analyze other columns.

In [11]:
df['ENROLL'] = df['ENROLL'].fillna(method='ffill')

In [12]:
df['ENROLL'] = df['ENROLL'].fillna(method='bfill')

In [ ]:
df.isnull().sum()

In [ ]:
df.groupby(['YEAR','STATE'])['TOTAL_REVENUE'].median()

In [48]:
# Dropping states with <= 23 value counts

# dropped_states = ['PUERTO_RICO', 'VIRGIN_ISLANDS','AMERICAN_SAMOA','NORTHERN_MARIANAS','DOD_DOMESTIC','DOD_OVERSEAS'
# ,'BUREAU_OF_INDIAN_AFFAIRS','BI','DD','NORTHERN_MARIANA_ISLANDS','COMMONWEALTH_OF_MARIANAS','DOD_-_OVERSEAS','BIE'
# ,'DOD_-_DOMESTIC','BUREAU_OF_INDIAN_EDUCATION','MARIANAS','DEPARTMENT_OF_DEFENSE','DEPARTMENT_OF_DEFENSE_EDUCATION_ACTIVITY'
# ,'AS','GU','PR','DOD','DOD_-_FOREIGN','DOD_(OVERSEAS_AND_DOMESTIC_COMBINED)','VI','MP','BUREAU_OF_INDIAN_EDUCATIO','U.S._VIRGIN_ISLANDS']                          1

df.drop(df[df['STATE'] == 'U.S._VIRGIN_ISLANDS'].index, inplace=True)

In [ ]:
#dropping YEAR 2017
df.drop(df[df['YEAR'] == 2017].index, inplace=True)

In [74]:
df.groupby('YEAR').agg({'OTHER_EXPENDITURE':'median'})

OTHER_EXPENDITURE
YEAR                   
1992            24913.0
1993           147419.0
1994           153540.0
1995           207663.5
1996           216519.5
1997           233118.5
1998           240608.0
1999           247541.0
2000           263027.5
2001           270376.5
2002           282522.0
2003           294360.0
2004           330735.5
2005           339343.0
2006           347621.0
2007           378230.0
2008           397140.0
2009           408292.0
2010           400773.0
2011           377766.5
2012           357434.5
2013           366989.0
2014           389309.5
2015           405674.5
2016           399899.0
2017           956321.0

In [ ]:
# replacing Other Expenditure missing values
df['OTHER_EXPENDITURE'].fillna(value = 248234.784314,inplace=True)


In [75]:
df.groupby('YEAR').agg({'OTHER_EXPENDITURE':'median'})

OTHER_EXPENDITURE
YEAR                   
1992            24913.0
1993           147419.0
1994           153540.0
1995           207663.5
1996           216519.5
1997           233118.5
1998           240608.0
1999           247541.0
2000           263027.5
2001           270376.5
2002           282522.0
2003           294360.0
2004           330735.5
2005           339343.0
2006           347621.0
2007           378230.0
2008           397140.0
2009           408292.0
2010           400773.0
2011           377766.5
2012           357434.5
2013           366989.0
2014           389309.5
2015           405674.5
2016           399899.0
2017           956321.0

In [ ]:
df['OTHER_EXPENDITURE'] = df['OTHER_EXPENDITURE'].fillna()

In [50]:
df.STATE.value_counts()

DISTRICT_OF_COLUMBIA    29
VIRGINIA                28
MAINE                   26
COLORADO                26
ARKANSAS                26
NEW_HAMPSHIRE           26
ALABAMA                 26
MINNESOTA               26
FLORIDA                 26
MICHIGAN                26
NEW_YORK                26
CALIFORNIA              26
ARIZONA                 26
RHODE_ISLAND            26
NEW_JERSEY              26
KENTUCKY                26
SOUTH_DAKOTA            26
MONTANA                 26
GEORGIA                 26
NORTH_DAKOTA            26
HAWAII                  26
DELAWARE                26
NEW_MEXICO              26
MISSOURI                26
NORTH_CAROLINA          26
INDIANA                 26
ILLINOIS                26
PENNSYLVANIA            26
CONNECTICUT             26
LOUISIANA               26
WEST_VIRGINIA           26
KANSAS                  26
WASHINGTON              26
MISSISSIPPI             26
NEBRASKA                26
OHIO                    26
TEXAS                   26
U

In [ ]:
# percentages of missing values
df.isnull().sum()*100/df.isnull().count()

In [49]:
df.isnull().sum()

PRIMARY_KEY                       0
STATE                             0
YEAR                              0
ENROLL                            0
TOTAL_REVENUE                    74
FEDERAL_REVENUE                  74
STATE_REVENUE                    74
LOCAL_REVENUE                    74
TOTAL_EXPENDITURE                74
INSTRUCTION_EXPENDITURE          74
SUPPORT_SERVICES_EXPENDITURE     74
OTHER_EXPENDITURE               125
CAPITAL_OUTLAY_EXPENDITURE       74
GRADES_PK_G                     118
GRADES_KG_G                     107
GRADES_4_G                      107
GRADES_8_G                      107
GRADES_12_G                     107
GRADES_1_8_G                    107
GRADES_9_12_G                   107
GRADES_ALL_G                    118
AVG_MATH_4_SCORE                818
AVG_MATH_8_SCORE                822
AVG_READING_4_SCORE             821
AVG_READING_8_SCORE             856
dtype: int64

In [ ]:
# finding percentages of missing values
df.isnull().sum()*100/df.isnull().count()

In [ ]:
df.groupby('YEAR').agg({'OTHER_EXPENDITURE':'mean'})

In [ ]:
df.columns

In [ ]:
df["AVG_MATH_4_SCORE"].fillna(df["AVG_MATH_4_SCORE"].mean(), inplace=True)
df["AVG_MATH_8_SCORE"].fillna(df["AVG_MATH_8_SCORE"].mean(), inplace=True)
df["AVG_READING_4_SCORE"].fillna(df["AVG_READING_4_SCORE"].mean(), inplace=True)
df["AVG_READING_8_SCORE"].fillna(df["AVG_READING_8_SCORE"].mean(), inplace=True)

In [13]:
# finding percentages of missing values
df.isnull().sum()*100/df.isnull().count()

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                           0.000000
TOTAL_REVENUE                   14.209115
FEDERAL_REVENUE                 14.209115
STATE_REVENUE                   14.209115
LOCAL_REVENUE                   14.209115
TOTAL_EXPENDITURE               14.209115
INSTRUCTION_EXPENDITURE         14.209115
SUPPORT_SERVICES_EXPENDITURE    14.209115
OTHER_EXPENDITURE               17.627346
CAPITAL_OUTLAY_EXPENDITURE      14.209115
GRADES_PK_G                     11.595174
GRADES_KG_G                      8.847185
GRADES_4_G                       8.780161
GRADES_8_G                       8.780161
GRADES_12_G                      8.780161
GRADES_1_8_G                     8.780161
GRADES_9_12_G                    8.780161
GRADES_ALL_G                    11.595174
AVG_MATH_4_SCORE                64.075067
AVG_MATH_8_SCORE                64.343164
AVG_READING_4_SCORE             64

In [ ]:
df.groupby(['YEAR','STATE'])['TOTAL_REVENUE'].median()

In [ ]:
# checking in any rows have all values missing
df.isnull().sum()

In [ ]:
df.shape

In [ ]:
# dropping any remaining minor null values
# df.dropna(how='all').shape

In [ ]:
df.columns

In [ ]:
# replacing missing values in GRADES columns with mean of that year

df['GRADES_PK_G'] = df.groupby('YEAR').GRADES_PK_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_KG_G'] = df.groupby('YEAR').GRADES_KG_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_4_G'] = df.groupby('YEAR').GRADES_4_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_8_G'] = df.groupby('YEAR').GRADES_8_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_1_8_G'] = df.groupby('YEAR').GRADES_1_8_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_9_12_G'] = df.groupby('YEAR').GRADES_9_12_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_ALL_G'] = df.groupby('YEAR').GRADES_ALL_G.transform(lambda x: x.fillna(x.mean()))
df['GRADES_12_G'] = df.groupby('YEAR').GRADES_12_G.transform(lambda x: x.fillna(x.mean()))

#'GRADES_KG_G','GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G','GRADES_9_12_G', 'GRADES_ALL_G'

In [ ]:
# checking in any rows have all values missing
df.isnull().sum()

Q3. This time, fill in the missing values using interpolation (extrapolation).

ANSWER:  We can interpolate for all missing values in the columns except that US territories or areas not part of the 50 US states should be removed because they are missing 64% of their values. For revenue and expenditures they're missing all values so it would be best to remove them. The remaining columns with missing values will be interpolated because the data is ordered by year or ordered by trends in time. 

In [14]:
df.columns

Index(['PRIMARY_KEY', 'STATE', 'YEAR', 'ENROLL', 'TOTAL_REVENUE',
       'FEDERAL_REVENUE', 'STATE_REVENUE', 'LOCAL_REVENUE',
       'TOTAL_EXPENDITURE', 'INSTRUCTION_EXPENDITURE',
       'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE',
       'CAPITAL_OUTLAY_EXPENDITURE', 'GRADES_PK_G', 'GRADES_KG_G',
       'GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G',
       'GRADES_9_12_G', 'GRADES_ALL_G', 'AVG_MATH_4_SCORE', 'AVG_MATH_8_SCORE',
       'AVG_READING_4_SCORE', 'AVG_READING_8_SCORE'],
      dtype='object')

In [16]:
# interpolating enrollment column
df["ENROLL"] = df["ENROLL"].interpolate(method='time')

In [51]:
df['ENROLL'].isnull().sum()

0

In [56]:
# interpolating revenue columns
df["TOTAL_REVENUE"] = df["TOTAL_REVENUE"].interpolate(method='ffill')
df["FEDERAL_REVENUE"] = df["FEDERAL_REVENUE"].interpolate(method='ffill')
df["STATE_REVENUE"] = df["STATE_REVENUE"].interpolate(method='ffill')
df["LOCAL_REVENUE"] = df["LOCAL_REVENUE"].interpolate(method='ffill')

In [57]:
df.isnull().sum()

PRIMARY_KEY                       0
STATE                             0
YEAR                              0
ENROLL                            0
TOTAL_REVENUE                     0
FEDERAL_REVENUE                   0
STATE_REVENUE                     0
LOCAL_REVENUE                     0
TOTAL_EXPENDITURE                74
INSTRUCTION_EXPENDITURE          74
SUPPORT_SERVICES_EXPENDITURE     74
OTHER_EXPENDITURE               125
CAPITAL_OUTLAY_EXPENDITURE       74
GRADES_PK_G                     118
GRADES_KG_G                     107
GRADES_4_G                      107
GRADES_8_G                      107
GRADES_12_G                     107
GRADES_1_8_G                    107
GRADES_9_12_G                   107
GRADES_ALL_G                    118
AVG_MATH_4_SCORE                818
AVG_MATH_8_SCORE                822
AVG_READING_4_SCORE             821
AVG_READING_8_SCORE             856
dtype: int64

In [62]:
# interpolating expenditure columns
df["TOTAL_EXPENDITURE"] = df["TOTAL_EXPENDITURE"].interpolate(method='ffill')
df["INSTRUCTION_EXPENDITURE"] = df["INSTRUCTION_EXPENDITURE"].interpolate(method='ffill')
df["SUPPORT_SERVICES_EXPENDITURE"] = df["SUPPORT_SERVICES_EXPENDITURE"].interpolate(method='ffill')
df["OTHER_EXPENDITURE"] = df["OTHER_EXPENDITURE"].interpolate(method='ffill')
df["CAPITAL_OUTLAY_EXPENDITURE"] = df["CAPITAL_OUTLAY_EXPENDITURE"].interpolate(method='ffill')

In [67]:
df.isnull().sum()

PRIMARY_KEY                       0
STATE                             0
YEAR                              0
ENROLL                            0
TOTAL_REVENUE                     0
FEDERAL_REVENUE                   0
STATE_REVENUE                     0
LOCAL_REVENUE                     0
TOTAL_EXPENDITURE                 0
INSTRUCTION_EXPENDITURE           0
SUPPORT_SERVICES_EXPENDITURE      0
OTHER_EXPENDITURE                 0
CAPITAL_OUTLAY_EXPENDITURE        0
GRADES_PK_G                       0
GRADES_KG_G                       0
GRADES_4_G                        0
GRADES_8_G                        0
GRADES_12_G                       0
GRADES_1_8_G                      0
GRADES_9_12_G                     0
GRADES_ALL_G                      0
AVG_MATH_4_SCORE                818
AVG_MATH_8_SCORE                822
AVG_READING_4_SCORE             821
AVG_READING_8_SCORE             856
dtype: int64

In [64]:
df["OTHER_EXPENDITURE"] = df["OTHER_EXPENDITURE"].interpolate(method='bfill')

In [66]:
# interpolating grades columns
df["GRADES_PK_G"] = df["GRADES_PK_G"].interpolate(method='ffill')
df["GRADES_KG_G"] = df["GRADES_KG_G"].interpolate(method='ffill')
df["GRADES_4_G"] = df["GRADES_4_G"].interpolate(method='ffill')
df["GRADES_8_G"] = df["GRADES_8_G"].interpolate(method='ffill')
df["GRADES_12_G"] = df["GRADES_12_G"].interpolate(method='ffill')
df["GRADES_1_8_G"] = df["GRADES_1_8_G"].interpolate(method='ffill')
df["GRADES_9_12_G"] = df["GRADES_9_12_G"].interpolate(method='ffill')
df["GRADES_ALL_G"] = df["GRADES_ALL_G"].interpolate(method='ffill')


In [72]:
df.isnull().sum()

PRIMARY_KEY                     0
STATE                           0
YEAR                            0
ENROLL                          0
TOTAL_REVENUE                   0
FEDERAL_REVENUE                 0
STATE_REVENUE                   0
LOCAL_REVENUE                   0
TOTAL_EXPENDITURE               0
INSTRUCTION_EXPENDITURE         0
SUPPORT_SERVICES_EXPENDITURE    0
OTHER_EXPENDITURE               0
CAPITAL_OUTLAY_EXPENDITURE      0
GRADES_PK_G                     0
GRADES_KG_G                     0
GRADES_4_G                      0
GRADES_8_G                      0
GRADES_12_G                     0
GRADES_1_8_G                    0
GRADES_9_12_G                   0
GRADES_ALL_G                    0
AVG_MATH_4_SCORE                0
AVG_MATH_8_SCORE                0
AVG_READING_4_SCORE             0
AVG_READING_8_SCORE             0
dtype: int64

In [71]:
# interpolating average grades columns
df["AVG_MATH_4_SCORE"] = df["AVG_MATH_4_SCORE"].interpolate(method='ffill')
df["AVG_MATH_8_SCORE"] = df["AVG_MATH_8_SCORE"].interpolate(method='ffill')
df["AVG_READING_4_SCORE"] = df["AVG_READING_4_SCORE"].interpolate(method='ffill')
df["AVG_READING_8_SCORE"] = df["AVG_READING_8_SCORE"].interpolate(method='bfill')

In [73]:
df.head(100)

PRIMARY_KEY                 STATE  YEAR     ENROLL  \
0                1992_ALABAMA               ALABAMA  1992    89711.0   
1                 1992_ALASKA                ALASKA  1992    89711.0   
2                1992_ARIZONA               ARIZONA  1992    89711.0   
3               1992_ARKANSAS              ARKANSAS  1992    89711.0   
4             1992_CALIFORNIA            CALIFORNIA  1992    89711.0   
5               1992_COLORADO              COLORADO  1992    89711.0   
6            1992_CONNECTICUT           CONNECTICUT  1992    89711.0   
7               1992_DELAWARE              DELAWARE  1992    89711.0   
8   1992_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1992    89711.0   
9                1992_FLORIDA               FLORIDA  1992    89711.0   
10               1992_GEORGIA               GEORGIA  1992    89711.0   
11                1992_HAWAII                HAWAII  1992    89711.0   
12                 1992_IDAHO                 IDAHO  1992    89711.0   
13              1992_ILLINOIS              ILLINOIS  1992    89711.0   
14               1992_INDIANA               INDIANA  1992    89711.0   
15                  1992_IOWA                  IOWA  1992    89711.0   
16                1992_KANSAS                KANSAS  1992    89711.0   
17              1992_KENTUCKY              KENTUCKY  1992    89711.0   
18             1992_LOUISIANA             LOUISIANA  1992    89711.0   
19                 1992_MAINE                 MAINE  1992    89711.0   
20              1992_MARYLAND              MARYLAND  1992    89711.0   
21         1992_MASSACHUSETTS         MASSACHUSETTS  1992    89711.0   
22              1992_MICHIGAN              MICHIGAN  1992    89711.0   
23             1992_MINNESOTA             MINNESOTA  1992    89711.0   
24           1992_MISSISSIPPI           MISSISSIPPI  1992    89711.0   
25              1992_MISSOURI              MISSOURI  1992    89711.0   
26               1992_MONTANA               MONTANA  1992    89711.0   
27              1992_NEBRASKA              NEBRASKA  1992    89711.0   
28                1992_NEVADA                NEVADA  1992    89711.0   
29         1992_NEW_HAMPSHIRE         NEW_HAMPSHIRE  1992    89711.0   
..                        ...                   ...   ...        ...   
70             1993_LOUISIANA             LOUISIANA  1993   795719.0   
71                 1993_MAINE                 MAINE  1993   211511.0   
72              1993_MARYLAND              MARYLAND  1993   751850.0   
73         1993_MASSACHUSETTS         MASSACHUSETTS  1993   862495.0   
74              2016_OKLAHOMA              OKLAHOMA  2016   672777.0   
75              1993_MICHIGAN              MICHIGAN  1993  1600650.0   
76             1993_MINNESOTA             MINNESOTA  1993   781566.0   
77           1993_MISSISSIPPI           MISSISSIPPI  1993   329408.0   
78              1993_MISSOURI              MISSOURI  1993   834636.0   
79               1993_MONTANA               MONTANA  1993   158875.0   
80              1993_NEBRASKA              NEBRASKA  1993   281354.0   
81                1993_NEVADA                NEVADA  1993   222846.0   
82         1993_NEW_HAMPSHIRE         NEW_HAMPSHIRE  1993   177722.0   
83            1993_NEW_JERSEY            NEW_JERSEY  1993   653488.0   
84            1993_NEW_MEXICO            NEW_MEXICO  1993   281407.0   
85              1993_NEW_YORK              NEW_YORK  1993  2663637.0   
86        1993_NORTH_CAROLINA        NORTH_CAROLINA  1993  1112636.0   
87          1993_NORTH_DAKOTA          NORTH_DAKOTA  1993   118793.0   
88                  1993_OHIO                  OHIO  1993  1834239.0   
89              1993_OKLAHOMA              OKLAHOMA  1993   312817.0   
90                1993_OREGON                OREGON  1993   509814.0   
91          1993_PENNSYLVANIA          PENNSYLVANIA  1993  1719829.0   
92          1993_RHODE_ISLAND          RHODE_ISLAND  1993   138939.0   
93        1993_SOUTH_CAROLINA        SOUTH_CAROLINA  1993 

In [76]:
df.describe()

YEAR        ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
count  1354.000000  1.354000e+03   1.354000e+03     1.354000e+03   
mean   2004.502216  9.182370e+05   9.830376e+06     8.122931e+05   
std       7.485215  1.039263e+06   1.187122e+07     1.131801e+06   
min    1992.000000  4.386600e+04   4.656500e+05     3.102000e+04   
25%    1998.000000  2.118078e+05   2.274300e+06     1.951748e+05   
50%    2005.000000  6.606715e+05   5.544502e+06     4.408520e+05   
75%    2011.000000  1.193205e+06   1.239841e+07     9.743288e+05   
max    2017.000000  6.307022e+06   8.921726e+07     9.990221e+06   

       STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
count   1.354000e+03   1.354000e+03       1.354000e+03   
mean    4.470618e+06   4.547465e+06       9.907614e+06   
std     5.505937e+06   5.665658e+06       1.205920e+07   
min     0.000000e+00   2.209300e+04       4.816650e+05   
25%     1.215930e+06   7.635780e+05       2.270249e+06   
50%     2.741092e+06   2.240796e+06       5.637836e+06   
75%     5.962061e+06   5.839841e+06       1.234903e+07   
max     5.090457e+07   3.610526e+07       8.532013e+07   

       INSTRUCTION_EXPENDITURE  SUPPORT_SERVICES_EXPENDITURE  \
count             1.354000e+03                  1.354000e+03   
mean              5.138865e+06                  2.916065e+06   
std               6.338252e+06                  3.420087e+06   
min               2.655490e+05                  1.399630e+05   
25%               1.238777e+06                  6.848360e+05   
50%               2.865617e+06                  1.636668e+06   
75%               6.300593e+06                  3.709028e+06   
max               4.396452e+07                  2.605802e+07   

       OTHER_EXPENDITURE         ...              GRADES_4_G     GRADES_8_G  \
count       1.354000e+03         ...             1354.000000    1354.000000   
mean        4.404789e+05         ...            64884.987445   64620.976366   
std         5.284729e+05         ...            80422.622299   79037.430575   
min         1.154100e+04         ...             2193.000000    1485.000000   
25%         9.403525e+04         ...            13642.250000   13228.250000   
50%         2.780130e+05         ...            43117.000000   43430.000000   
75%         5.638142e+05         ...            75690.750000   77119.750000   
max         3.995951e+06         ...           493415.000000  500143.000000   

         GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
count    1354.000000  1.354000e+03   1.354000e+03  1.354000e+03   
mean    54677.895126  5.220470e+05   2.487779e+05  7.891371e+05   
std     67965.993813  6.446954e+05   3.080847e+05  9.645945e+05   
min       484.000000  1.819600e+04   2.758000e+03  2.455400e+04   
25%     10853.500000  1.094928e+05   5.083575e+04  1.694818e+05   
50%     37319.500000  3.446885e+05   1.701055e+05  5.348650e+05   
75%     67632.250000  6.133735e+05   2.918508e+05  9.196878e+05   
max    498403.000000  3.929869e+06   2.013687e+06  5.944746e+06   

       AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
count       1354.000000       1354.000000          1354.000000   
mean         234.389232        279.102824           220.682560   
std            9.578056          7.832439             5.777825   
min          187.134670        232.831510           178.557612   
25%          225.377592        275.080076           219.010780   
50%          237.326270        277.472348           221.185556   
75%          242.542822        286.104034           223.619939   
max          253.420961        300.568235           236.773867   

       AVG_READING_8_SCORE  
count          1354.000000  
mean            262.568092  
std               6.122645  
min             236.379102  
25%             258.108365  
50%             262.118986  
75%             267.586767  
max             280.499130  

[8 rows x 23 columns]

Filling in missing enrollments using interpolation seemed to detect the temporal trends in the data much more than using the mean for the year, however, values tended to be more extreme max and mins. Values for more of the missing data in the average columns tended to be higher as well but more in line with trends by year. 